### ISSUE TITLE
deepcopy exception is a breaking change

### ISSUE THREAD

**Issue by michael-123123** [cvxpy/cvxpy#2012]
**Describe the bug**
PR #1852 adds an override to `__deepcopy__` in cvxpy objects, 
which now throws an exception instead of using the default implementation provided by `object`.
Our code base contains multiple classes which have cvxpy objects as attributes.
These classes contain data as well as cvxpy models, and we `deepcopy` them.

For example, we might have a list which is partially numbers (or cvxpy constants) and partially cvxpy variables.
And they may be deepcopied (usually after the model is solved) as part of a serialization process.
This worked fine until the changes in PR #1852

It is not realistic for us to customize the entire class hierarchy to call `.copy` for cvxpy objects whenever `deepcopy` is used.
I would suggest that either:
1. `__deepcopy__` will not throw an exception and use the default implementation or
2. `__deepcopy__` will print a warning and call `.copy` instead

**To Reproduce**
```
expr = cp.Variable()
deepcopy(expr)
```

**Expected behavior**
A deep copy of the expression, whatever that means semantically.

**Version**
 - OS: Windows/Linux 
 - CVXPY Version: 1.3.0

**Comment by adishavit** [comment]
Indeed, this is breaking our code.  
Even after reviewing the PR #1852 I don't understand:

1. The original semantics of `copy` and `deepcopy`;
2. The indented semantics for `copy` and `deepcopy` ;
3. How the PR fixes the original semantics - and if the change is worth the deep breaking changes it entails.

**Comment by phschiele** [comment]
@michael-123123 @adishavit Thanks for reporting this.

The intended semantics are that if you create a copy of a CVXPY expression (as originally intended via the `.copy()` method) the copied expression still refers to the same instance of the variables.

The PR was initially meant to fix a [subtle bug](https://github.com/cvxpy/cvxpy/issues/1827#issuecomment-1184870476), but after some discussion, I was questioning the reasoning for when a deepcopy of a cvxpy expression even makes sense. It turned out that that consideration was too narrow, as, for example, deep copying a mutable object that contains CVXPY expressions as well as other objects, can make sense. This seems to be the use case reported above. Indeed, found out as much myself here, where I had to resort to an inelegant workaround: https://github.com/robertmartin8/PyPortfolioOpt/pull/493/files). 

I would support restoring the deepcopy functionality with the soon-to-be-released 1.3.1 patch, if there are no objections by @cvxpy/maintainers.

**Comment by SteveDiamond** [comment]
Agreed, let's restore the default deep copy behavior.

**Comment by SteveDiamond** [comment]
Let's also add a test reflecting the described use case.

**Comment by adishavit** [comment]
I think it would also be advantageous to explicitly document the desired
semantics and the expected behavior of the (reverted, default) copy and
deepcopy.

I’m still not sure what it means to have eg two distinct Variables with the
same `id`.

**Comment by michael-123123** [comment]
I think a good argument could be made that a deepcopy of a cvxpy expression should instantiate a new expression tree with fresh variables, maintaining the same relationships as their original counterparts. This is not exactly object's default behavior and not exactly `.copy`'s behavior, so perhaps it's out of scope.

Notice that in either case, reverting to the old behavior (which I'm all for) still leaves some bugs.
For example (ran on cvxpy 1.2.1):
```
>>> a = Variable()
>>> b = deepcopy(a)
>>> c = [a >= 0, b >= 0, a+b <= 10, a >= b + 1]
>>> p = Problem(Maximize(a+b), c)
>>> p.solve()
-inf
>>> p.status
'infeasible'
```

**Comment by SteveDiamond** [comment]
> I think a good argument could be made that a `deepcopy` of a cvxpy expression should instantiate a new expression tree with fresh variables, maintaining the same relationships as their original counterparts. This is not exactly `object`'s default behavior and not exactly `.copy`'s behavior, so perhaps it's out of scope.
> 
> Notice that in either case, reverting to the old behavior (which I'm all for) still leaves some bugs. For example (ran on cvxpy 1.2.1):
> 
> ```
> >>> a = Variable()
> >>> b = deepcopy(a)
> >>> c = [a >= 0, b >= 0, a+b <= 10, a >= b + 1]
> >>> p = Problem(Maximize(a+b), c)
> >>> p.solve()
> -inf
> >>> p.status
> 'infeasible'
> ```

I don't see this behavior as a bug. If you copy or deepcopy ``a``, regardles, it's just the same variable with a different name. So ``a >= b + 1`` is infeasible.

@adishavit in regards to your question about semantics, a variable is defined by its id. If two variables have the same id they are aliases, equivalent to writing ``var1 = var2``.

**Comment by SteveDiamond** [comment]
> Agreed, let's restore the default deep copy behavior.

I've reviewed this issue more and I now think that deepcopy should be the same as copy. I'm not sure if that's the default behavior of deepcopy or not, but the cvxpy copy semantics are well defined (if not well documented).

**Comment by rileyjmurray** [comment]
Quick chime-in (I'm on a plane at the moment and pretty sleep deprived so I don't have a great comment at this point):

Let's think carefully about getting the semantics of this right and clearly documenting them. I'd personally want a clear statement of different options and pros and cons laid out for each. This could be done in a Google doc if others would be okay with that.

**Comment by SteveDiamond** [comment]
Fixed in 1.3.1
